src/client/elements/pageNavNewAPI.ml

```diff
@@ -0,0 +1,157 @@
+open Nes
+open Dancelor_common_model
+open Dancelor_client_html.NewAPI
+
+type state = {
+  current_page : int; (* first page is [1] *)
+  entries_per_page : int;
+  number_of_entries : int;
+}
+
+let number_of_pages state =
+  (state.number_of_entries
+   + state.entries_per_page
+   - 1)
+  / state.entries_per_page
+
+let current_pagination state =
+  (* NOTE: Our page numbers start at 1. *)
+  Pagination.{
+    start = (state.current_page - 1) * state.entries_per_page;
+    end_ = min (state.current_page * state.entries_per_page) state.number_of_entries;
+  }
+
+type t = {
+  signal : state React.signal;
+  setter : state -> unit;
+  updater : (state -> state) -> unit;
+}
+
+let create ~number_of_entries ~entries_per_page =
+  let initial = {
+    current_page = 1;
+    entries_per_page;
+    number_of_entries = 0;
+  }
+  in
+  let (signal, setter) = S.create initial in
+  let updater f = setter (f (S.value signal)) in
+  Lwt.on_success number_of_entries (fun number_of_entries ->
+      updater (fun state -> { state with number_of_entries }));
+  { signal; setter; updater }
+
+let status_text pagination =
+  flip S.map pagination.signal @@ fun state ->
+  if state.number_of_entries = 0 then
+    "No entries"
+  else
+    let pagination = current_pagination state in
+    spf "Showing %i to %i of %i entries"
+      (Pagination.start pagination + 1)
+      (Pagination.end_ pagination)
+      state.number_of_entries
+
+module Button = struct
+  (** [make ~active ~enabled ~target ~text pagination] is a button that shows
+      [text]. It is active (that is, it shows as different from the other) when
+      the predicate [active] returns [true] on the [pagination] state. It is
+      enabled (that is not grayed out and clickable) when the predicate [enabled]
+      returns [true] on the [pagination] state. It changes the page to the result
+      of [target] applied on the [pagination] state. *)
+  let make ~active ~enabled ~target ~text pagination =
+    li
+      ~a:[
+        R.a_class (
+          flip S.map pagination.signal @@ fun state ->
+          if active state then ["active"] else []
+        )
+      ]
+      [
+        button ~a: [
+          a_button_type `Button;
+          R.a_class (
+            flip S.map pagination.signal @@ fun state ->
+            if enabled state then ["clickable"] else ["disabled"]
+          );
+          a_onclick
+            (fun _ ->
+               let state = S.value pagination.signal in
+               if enabled state then
+                 pagination.updater (fun state ->
+                     { state with current_page = target state }
+                   );
+               false
+            )
+        ]
+          [txt text]
+      ]
+
+  (** A button that is never enabled and shows three dots. *)
+  let ellipsis =
+    make
+      ~active:(Fun.const false)
+      ~enabled:(Fun.const false)
+      ~target:(fun _ -> assert false)
```

Comment:
I'd be in favor of making this target some kind of named constant, to avoid duplicating "assert false" for other buttons (if any) that do not redirect.

Reply:
So you'd want a `val no_target : state -> int` that would basically be `let no_target _ = assert false`?

Reply:
(I'm down; I'm just not sure I understood properly.)

Reply:
Very good point :p Done in e2fdabc3767439b49b599ae8c41e89994fcf479d.